### DISCUSSION TITLE
sktime with Tbats problem with forecasting horizon

### DISCUSSION THREAD

**Discussion by Nikhitha74** [alan-turing-institute/sktime#4145] [General]
Hi,
 
I have another doubt regarding sktime with tbats. There is a problem with the forecasting horizon i give. Reproducing the code with dummy data below:

import pandas as pd
import numpy as np
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.bats import BATS
from sktime.forecasting.tbats import TBATS

#get the data
data = {'Date': ['2000-01-03', '2000-01-04', '2000-01-05', '2000-01-06',
                 '2000-01-07','2000-01-10','2000-01-11','2000-01-12',
                 '2000-01-13','2000-01-14','2000-01-17'],
        'open': [694,714,716.25,741,832.4,843.15,902,905,900,879.9,904.95],
        'high':[708.5,729,758.9,784.3,847.05,847.05,952,924,936,903.7,904.95],
        'close':[708.5,712.35,726.2,784.3,847.05,847.05,883.7,888.35,879.75,
                 887.55,817.35]
      }
# Create DataFrame
df = pd.DataFrame(data)
  
# Print the datatypes of output.
df.dtypes

#Date variable
df.Date = pd.to_datetime(df.Date,format = "%Y-%m-%d")
df1 = df.set_index('Date')
df1.head(2)

y = df1.drop(columns=["open", "high"])
X=df1.drop(columns=["close"])
X.head(3)

y_train, y_test, X_train, X_test = temporal_train_test_split(y, X, test_size=4)
forecaster = TBATS(use_trend=False, 
                   use_box_cox=False)
forecaster.fit(y=y_train,X=X_train)

# Forecasting horizon, same as test data
fh1 = ForecastingHorizon(y_test.index, is_relative=False)

y_pred = forecaster.predict(X=X_test,fh=fh1)#For test set

Error is thrown from the above line of code

C:\Users\Nikhitha\anaconda3\envs\for_Sktime\lib\site-packages\sktime\forecasting\base\_fh.py:802: FutureWarning: Timestamp.freq is deprecated and will be removed in a future version.
  freq = x.freq
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
Cell In[12], line 1
----> 1 y_pred = forecaster.predict(X=X_test,fh=fh1)

File ~\anaconda3\envs\for_Sktime\lib\site-packages\sktime\forecasting\base\_base.py:404, in BaseForecaster.predict(self, fh, X)
    402 # we call the ordinary _predict if no looping/vectorization needed
    403 if not self._is_vectorized:
--> 404     y_pred = self._predict(fh=fh, X=X_inner)
    405 else:
    406     # otherwise we call the vectorized version of predict
    407     y_pred = self._vectorize("predict", X=X_inner, fh=fh)

File ~\anaconda3\envs\for_Sktime\lib\site-packages\sktime\forecasting\base\adapters\_tbats.py:152, in _TbatsAdapter._predict(self, fh, X)
    137 def _predict(self, fh, X=None):
    138     """Forecast time series at future horizon.
    139 
    140     Parameters
   (...)
    150         Point predictions
    151     """
--> 152     return self._tbats_forecast(fh)

File ~\anaconda3\envs\for_Sktime\lib\site-packages\sktime\forecasting\base\adapters\_tbats.py:167, in _TbatsAdapter._tbats_forecast(self, fh)
    154 def _tbats_forecast(self, fh):
    155     """TBATS forecast without confidence interval.
    156 
    157     Parameters
   (...)
    165         Prediction
    166     """
--> 167     fh = fh.to_relative(cutoff=self.cutoff)
    169     if not fh.is_all_in_sample(cutoff=self.cutoff):
    170         fh_out = fh.to_out_of_sample(cutoff=self.cutoff)

File ~\anaconda3\envs\for_Sktime\lib\site-packages\sktime\forecasting\base\_fh.py:461, in ForecastingHorizon.to_relative(self, cutoff)
    446 """Return forecasting horizon values relative to a cutoff.
    447 
    448 Parameters
   (...)
    458     Relative representation of forecasting horizon.
    459 """
    460 cutoff = self._coerce_cutoff_to_index_element(cutoff)
--> 461 return _to_relative(fh=self, cutoff=cutoff)

File ~\anaconda3\envs\for_Sktime\lib\site-packages\sktime\forecasting\base\_fh.py:688, in _to_relative(fh, cutoff)
    684 if isinstance(absolute, pd.DatetimeIndex):
    685     # coerce to pd.Period for reliable arithmetics and computations of
    686     # time deltas
    687     absolute = _coerce_to_period(absolute, freq=fh.freq)
--> 688     cutoff = _coerce_to_period(cutoff, freq=fh.freq)
    690 # TODO: Replace when we upgrade our lower pandas bound
    691 #  to a version where this is fixed
    692 # Compute relative values
   (...)
    704 # Once the bug is fixed the line should simply be:
    705 # relative = absolute - cutoff
    706 relative = pd.Index([date - cutoff for date in absolute])

File ~\anaconda3\envs\for_Sktime\lib\site-packages\sktime\forecasting\base\_fh.py:803, in _coerce_to_period(x, freq)
    801 if isinstance(x, pd.Timestamp) and freq is None:
    802     freq = x.freq
--> 803     raise ValueError(
    804         "_coerce_to_period requires freq argument to be passed if x is pd.Timestamp"
    805     )
    806 try:
    807     return x.to_period(freq)

ValueError: _coerce_to_period requires freq argument to be passed if x is pd.Timestamp

Could someone let me know how to rectify this.

**Comment by fkiraly** [comment]
getting the code in one cell:

```python
import pandas as pd
import numpy as np
from sktime.forecasting.model_selection import temporal_train_test_split
from sktime.forecasting.base import ForecastingHorizon
from sktime.forecasting.bats import BATS
from sktime.forecasting.tbats import TBATS

#get the data
data = {'Date': ['2000-01-03', '2000-01-04', '2000-01-05', '2000-01-06',
'2000-01-07','2000-01-10','2000-01-11','2000-01-12',
'2000-01-13','2000-01-14','2000-01-17'],
'open': [694,714,716.25,741,832.4,843.15,902,905,900,879.9,904.95],
'high':[708.5,729,758.9,784.3,847.05,847.05,952,924,936,903.7,904.95],
'close':[708.5,712.35,726.2,784.3,847.05,847.05,883.7,888.35,879.75,
887.55,817.35]
}

df = pd.DataFrame(data)

#Date variable
df.Date = pd.to_datetime(df.Date,format = "%Y-%m-%d")
df1 = df.set_index('Date')
df1.head(2)

y = df1.drop(columns=["open", "high"])
X=df1.drop(columns=["close"])
X.head(3)

y_train, y_test, X_train, X_test = temporal_train_test_split(y, X, test_size=4)
forecaster = TBATS(use_trend=False,
use_box_cox=False)
forecaster.fit(y=y_train,X=X_train)

fh1 = ForecastingHorizon(y_test.index, is_relative=False)

y_pred = forecaster.predict(X=X_test,fh=fh1)#For test set
```

**Comment by fkiraly** [comment]
looks like a bug, moved to an issue here: https://github.com/sktime/sktime/issues/4151

**Reply by Nikhitha74** [reply]
Hi fkiraly,
Thank you for the response. Will this be sorted out soon? Asking so that I can use tbats.
